src/modules/control_allocator/ControlAllocator.cpp

```diff
@@ -460,6 +474,168 @@ ControlAllocator::Run()
 	perf_end(_loop_perf);
 }
 
+void ControlAllocator::preflight_check_handle_command()
+{
+	vehicle_command_s vehicle_command;
+
+	if (_vehicle_command_sub.update(&vehicle_command)) {
+
+		uint8_t result = vehicle_command_ack_s::VEHICLE_CMD_RESULT_ACCEPTED;
+
+		if (vehicle_command.command == vehicle_command_s::VEHICLE_CMD_DO_PREFLIGHT_CS_CHECK) {
+			if (!_armed) {
+				// currently this does not check prearmed status. if not prearmed, it will just do nothing.
+				// should we output some sort of mild warning in that case?
+
+				preflight_check_start(vehicle_command);
+				result = vehicle_command_ack_s::VEHICLE_CMD_RESULT_IN_PROGRESS;
+
+			} else {
+				result = vehicle_command_ack_s::VEHICLE_CMD_RESULT_TEMPORARILY_REJECTED;
+				PX4_INFO("Control surface preflight check rejected (armed)");
+			}
+
+			preflight_check_send_ack(result);
+		}
+	}
+}
+
+void ControlAllocator::preflight_check_start(vehicle_command_s &cmd)
+{
+
+	// If one is running, abort it. Depending on the use case we may prefer
+	// to instead "silently" overwrite the value.
+	if (_preflight_check_running) {
+		preflight_check_abort();
+	}
+
+	int axis = (uint8_t) lroundf(cmd.param1);
+	float input = cmd.param2;
+
+	_preflight_check_running = true;
+	_preflight_check_axis = axis;
+	_preflight_check_input = input;
+	_preflight_check_started = hrt_absolute_time();
```

Comment:
Pass the time now from the outside (add it as an argument of `preflight_check_start()`. The allocator runs very fast and `hrt_absolute_time()` calls are expensive.